x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/CacheInvalidatorRegistry.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.security.support;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static org.elasticsearch.xpack.security.support.SecurityIndexManager.isIndexDeleted;
+import static org.elasticsearch.xpack.security.support.SecurityIndexManager.isMoveFromRedToNonRed;
+
+/**
+ * A registry that provides common cache invalidation services for caches that relies on the security index.
+ */
+public class CacheInvalidatorRegistry {
+
+    private final Map<String, CacheInvalidator> cacheInvalidators = new ConcurrentHashMap<>();
+
+    public CacheInvalidatorRegistry() {
+    }
+
+    public void registerCacheInvalidator(String name, CacheInvalidator cacheInvalidator) {
+        if (cacheInvalidators.containsKey(name)) {
+            throw new IllegalArgumentException("Cache invalidator registry already has an entry with name: [" + name + "]");
+        }
+        cacheInvalidators.put(name, cacheInvalidator);
+    }
+
+    public void onSecurityIndexStageChange(SecurityIndexManager.State previousState, SecurityIndexManager.State currentState) {
+        if (isMoveFromRedToNonRed(previousState, currentState)
+            || isIndexDeleted(previousState, currentState)
+            || previousState.isIndexUpToDate != currentState.isIndexUpToDate) {
+            cacheInvalidators.values().forEach(CacheInvalidator::invalidateAll);
+        }
+    }
+
+    public void invalidateByKey(String cacheName, Collection<String> keys) {
+        final CacheInvalidator cacheInvalidator = cacheInvalidators.get(cacheName);
+        if (cacheInvalidator != null) {
+            cacheInvalidator.invalidate(keys);
+        }
```

Comment:
The lack of an `else` seems like dangerous leniency. Why do we silently pretend to invalidate a cache that doesn't exist?

Reply:
This is a good point and there is no reason for this leniency. Add an [else branch](https://github.com/elastic/elasticsearch/pull/59376/commits/ef95a1dd11410530bccba9d0ddbbb46d50ea4e82#diff-95381cf595a43e010005cf0de5538f56) to throw exception and also relevant tests.